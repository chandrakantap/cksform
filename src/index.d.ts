export as namespace CKSForm;

interface iCKSForm {
    handleSubmit: Function,
    onChangeField: Function,
    resetForm: Function,
    validateForm: Function,
    validateField: Function,
    errors: object,
    setErrors: Functions,
}


declare function useCKSForm(defaultValue: object): iCKSForm;
export = useCKSForm;