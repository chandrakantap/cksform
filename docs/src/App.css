*{
  margin: 0;
  padding: 0;
  box-sizing: border-box;
}
body{
  background: #DADADA;
  overflow: hidden;
}
.App {
    margin: 0 auto;
    padding: 2rem;
    height: calc(100vh);
    overflow-x: hidden;
    overflow-y: auto;
    max-width:800px;
    font-family: Arial, Helvetica, sans-serif;
    background: #FFF;
  }
h1{
  font-size: 1.5rem;
  font-weight: 400;
  margin-bottom: 1rem;
}
form{
  display: grid;
  grid-gap:10px;
  grid-template-columns: 1fr 2fr 1fr 2fr;
  align-items: baseline;
  margin-top: 4rem;
}
label{
  font-size: 1.2rem;
  text-align:right;
}
input{
  width: 100%;
  padding: 0.5rem;
  font-size: 16px;
  font-family: inherit;
}
input:error{
  border-color: red;
}
input[type='button'],button{
  padding: 0.5rem 1rem;
  width: max-content;
  cursor: pointer;
}
ul.error{
  font-size:0.8rem;
  color:red;
  margin-top:5px;
  list-style-type: none;
}
.btnLbl{
  padding: 5px;
}