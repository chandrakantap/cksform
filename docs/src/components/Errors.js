import React from 'react';

export default function ({ errors = {}, fieldName }) {
    const fieldErrors = errors[fieldName];
    if (!Array.isArray(fieldErrors) || fieldErrors.length === 0) {
        return null;
    }
    return (
      <ul className="error">
        {fieldErrors.map((error, index) => (<li key={index}>{error}</li>))}
      </ul>
    )
}