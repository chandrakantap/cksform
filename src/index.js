/* eslint-disable @typescript-eslint/explicit-function-return-type */

import { useState } from 'react';

const getErrors = (element) => {
  const elementDisplayName = element.dataset.displayName || element.name;
  const errors = [];
  const invalidValueEm = `Invalid value provided for ${elementDisplayName}.`;

  if (element.validity.patternMismatch) {
    errors.push(element.dataset.emPatternMismatch || invalidValueEm);
  }
  if (element.validity.rangeOverflow) {
    errors.push(element.dataset.emRangeOverflow || `${elementDisplayName} must be below ${element.max}.`);
  } else if (element.validity.rangeUnderflow) {
    errors.push(element.dataset.emRangeOverflow || `${elementDisplayName} must be atleast ${element.min}.`);
  }

  if (element.validity.stepMismatch) {
    errors.push(element.dataset.emStepMismatch || invalidValueEm);
  }

  if (element.validity.valueMissing) {
    errors.push(element.dataset.emValueMissing || `${elementDisplayName} is required.`);
  }
  if (element.validity.typeMismatch) {
    errors.push(element.dataset.emTypeMismatch || `Please provide a valid ${element.type}.`);
  }

  if (element.validity.tooShort) {
    errors.push(element.dataset.emTooShort || `Length must be atleast ${element.minLength}.`);
  } else if (element.validity.tooLong) {
    errors.push(element.dataset.emTooLong || `Length should not be more than ${element.maxLength}.`);
  }

  if (errors.length === 0 && !element.validity.valid) {
    errors.push('Invalid data.')
  }
  return errors;
};


const useCKSForm = (initialValue = {}) => {
  const [formData, setFormData] = useState({ ...initialValue });
  const [errors, setErrors] = useState({});

  const resetForm = () => setFormData({ ...initialValue });

  const handleSubmit = (callBack) => (event) => {
    const { form } = event.target;
    const isFormValid = form.checkValidity();
    const formErrors = isFormValid
      ? undefined
      : [...form.elements]
        .filter((element) => Boolean(element.name))
        .filter((element) => !element.checkValidity())
        .reduce(
          (elementErrors, element) => ({
            ...elementErrors,
            [element.name]: getErrors(element),
          }),
          {},
        );

    if (formErrors) {
      setErrors(formErrors)
    } else {
      setErrors({});
      callBack && callBack(event, formData);
    }

  };

  const validateField = (event) => {
    if (event.target.name) {
      const element = event.target;
      const formErrors = { ...errors };
      if (element.validity.valid) {
        formErrors[element.name] = undefined;
      } else {
        formErrors[element.name] = getErrors(element);
      }
      setErrors(formErrors);
    }
  };

  const onChangeField = (event) => {
    if (event.target.name) {
      const targetType = (event.target.type || 'UNKNOWN').toUpperCase();
      if (targetType === 'CHECKBOX') {
        const { checked } = event.target;
        const nextValue = checked
          ? [...formData[event.target.name], event.target.value]
          : formData[event.target.name].filter(
            (val) => val !== event.target.value,
          );
        setFormData({ ...formData, [event.target.name]: nextValue });
      } else {
        setFormData({ ...formData, [event.target.name]: event.target.value });
      }
    }
  };

  return {
    formData,
    onChangeField,
    resetForm,
    handleSubmit,
    validateField,
    errors,
    setErrors,
  };
};

export default useCKSForm;
