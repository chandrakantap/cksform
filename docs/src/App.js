import React, { Component } from "react";
import { hot } from "react-hot-loader";
import RegistrationForm from './components/RegistrationForm';
import "./App.css";

class App extends Component {
    render() {
        return (
            <div className="App">
                <h1>Simple Form</h1>
                <RegistrationForm />
            </div>
        );
    }
}

export default hot(module)(App);