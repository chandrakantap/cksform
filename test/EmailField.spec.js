import React from 'react';
import { mount } from 'enzyme';
import { createSerializer } from 'enzyme-to-json';
import { useCKSForm } from '../src/index';
import Errors from '../docs/src/components/Errors';

expect.addSnapshotSerializer(createSerializer({ mode: 'deep', noKey: true, ignoreDefaultProps: true }));

const EmailFieldForm = ({ onSubmit, ...props }) => {
  const { formData, onChangeField, errors, handleSubmit, validateField } = useCKSForm({ emailId: '' });
  return (
    <form noValidate autoComplete="off">
      <input
        type="email"
        name="emailId"
        required
        minLength={10}
        maxLength={12}
        value={formData.emailId}
        onChange={onChangeField}
        onBlur={validateField}
        data-display-name="Email ID"
        {...props}
      />
      <Errors errors={errors} fieldName="emailId" />
      <input id="submitBtn" type="button" value="Submit" onClick={handleSubmit(onSubmit)} />
    </form>
  );
}
describe('Email field', () => {
  test('renders properly', () => {
    const warapper = mount(<EmailFieldForm />);
    expect(warapper).toMatchSnapshot();
  });

  test('Validates when no value provided', () => {
    const warapper = mount(<EmailFieldForm />);
    warapper.find("input[type='button']").first().simulate('click');
    expect(warapper).toMatchSnapshot();
  });

  test('Validates when invalid value provided', () => {
    const warapper = mount(<EmailFieldForm />);
    warapper.find("input[type='email']").first().simulate('change', { target: { name: 'emailId', value: "pal.chand" } });
    warapper.find("input[type='button']").first().simulate('click');
    expect(warapper).toMatchSnapshot();
  });

  test('Call onSubmit if validation successful', () => {
    const onSubmit = jest.fn();
    const warapper = mount(<EmailFieldForm onSubmit={onSubmit} />);
    warapper.find("input[type='email']").first().simulate('change', { target: { name: 'emailId', value: "abcd@abc.com" } });
    warapper.find("input[type='button']").first().simulate('click');
    expect(onSubmit).toHaveBeenCalled();
    expect(warapper).toMatchSnapshot();
  });
})
