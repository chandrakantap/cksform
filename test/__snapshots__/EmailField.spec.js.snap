// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Email field Call onSubmit if validation successful 1`] = `
<form
  autoComplete="off"
  noValidate={true}
>
  <input
    data-display-name="Email ID"
    maxLength={12}
    minLength={10}
    name="emailId"
    onBlur={[Function]}
    onChange={[Function]}
    required={true}
    type="email"
    value="abcd@abc.com"
  />
  <input
    id="submitBtn"
    onClick={[Function]}
    type="button"
    value="Submit"
  />
</form>
`;

exports[`Email field Validates when invalid value provided 1`] = `
<form
  autoComplete="off"
  noValidate={true}
>
  <input
    data-display-name="Email ID"
    maxLength={12}
    minLength={10}
    name="emailId"
    onBlur={[Function]}
    onChange={[Function]}
    required={true}
    type="email"
    value="pal.chand"
  />
  <ul
    className="error"
  >
    <li>
      Please provide a valid email.
    </li>
  </ul>
  <input
    id="submitBtn"
    onClick={[Function]}
    type="button"
    value="Submit"
  />
</form>
`;

exports[`Email field Validates when no value provided 1`] = `
<form
  autoComplete="off"
  noValidate={true}
>
  <input
    data-display-name="Email ID"
    maxLength={12}
    minLength={10}
    name="emailId"
    onBlur={[Function]}
    onChange={[Function]}
    required={true}
    type="email"
    value=""
  />
  <ul
    className="error"
  >
    <li>
      Email ID is required.
    </li>
  </ul>
  <input
    id="submitBtn"
    onClick={[Function]}
    type="button"
    value="Submit"
  />
</form>
`;

exports[`Email field renders properly 1`] = `
<form
  autoComplete="off"
  noValidate={true}
>
  <input
    data-display-name="Email ID"
    maxLength={12}
    minLength={10}
    name="emailId"
    onBlur={[Function]}
    onChange={[Function]}
    required={true}
    type="email"
    value=""
  />
  <input
    id="submitBtn"
    onClick={[Function]}
    type="button"
    value="Submit"
  />
</form>
`;
